schema/spacewalk/postgres/tables/rhnPackageCapability_index.sql

```diff
@@ -13,13 +13,13 @@
 -- in this software or its documentation.
 --
 
-CREATE UNIQUE INDEX rhn_pkg_cap_name_version_uq
-    ON rhnPackageCapability (name, version)
- where version is not null;
+CREATE INDEX rhn_pkg_cap_name_version_uq
+    ON rhnPackageCapability USING btree (sha512(replace(name, '\', '\\')::bytea), version)
```

Comment:
What is the reason for the indexing of the column hash instead of the column?

Reply:
In case of using UNIQUE only `btree` can be used, but in case of large values it could lead to the following issue:
```
2021/08/03 14:41:21 +03:00 index row size 3488 exceeds btree version 4 maximum 2704 for index "rhn_pkg_cap_name_uq"
DETAIL:  Index row references tuple (215928,89) in relation "rhnpackagecapability".
HINT:  Values larger than 1/3 of a buffer page cannot be indexed.
Consider a function index of an MD5 hash of the value, or use full text indexing.
```
Using the hash is the recommended solution for such cases.
Will do some extra checks with the speed of reindexing.

Reply:
I did some more tests to check index creation speed and reindex and it seems reindexing with hash runs almost twice faster (I tested it with Uyuni installed on Raspberry Pi 4), but as I said before full text indexing is one of the recommendations, but it can't be used with `btree` which is required for `UNIQUE`. MD5 is not reliable, sha512 used instead, but it requres additionally replace `/` with `//`

Reply:
There are two sides to consider here:
 1- the `UNIQUE` part - this guarantees there are no duplicates. From that perspective, taking the hash of values and index hashes works (although it is a bit redundant, one could in principle use a hash index instead of btree of hashes - but that's unimportant)
 2- the `INDEX` part - this is about speeding up queries that search by a certain column. If you apply a function to the indexed column (`sha512(...)` in this case), then _only queries which search by that function will get performance improvements_.

IOW:

```sql
CREATE INDEX fast_index ON table (column);
```

speeds up queries like:
```sql
SELECT * FROM table WHERE colum = 'value';
```

While

```sql
CREATE INDEX fast_index ON table (sha512(column));
```

**only** speeds up queries like:
```sql
SELECT * FROM table WHERE column = sha512('value');
```

So queries like:
```sql
SELECT * FROM table WHERE column = 'value';
```

will **not** benefit from the index.


In this case, I fear we do have queries looking up capabilities by their name around the codebase, can you please check?

Reply:
Just finished the test of uploading `16 542 013` records (`15 530 913` uniq records) to test tables with the same stracture and indexes as `rhnPackageCapability` has now, and with modified table. It takes `38m37.023s` minutes to upload all the data with old table structure and indexes and `28m6.253s` for modified table. Upload was made the same way as it's done in `reposync`. No any side effect from the IO, the tests were performed on NVMe drive. I think full text index helps for cases like:
```
SELECT * FROM table WHERE column = 'value';
```
as Postgres using indexes in this case the proper way by getting the hash and comparing with the indexes it has.

Reply:
Full text index is for searching by keyword, ie. for natural language and search-engine kinds of queries, which is not what we are looking for here.

Problem is in any code piece that `SELECT`s `rhnPackageCapability` by `name`, or `ORDER BY name` or `JOIN` (potentially). I did a quick search and could find the following:

https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/common/db/datasource/xml/Package_queries.xml#L138-L274
https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/domain/rhnpackage/Package_satellite.hbm.xml#L237-L263
https://github.com/uyuni-project/uyuni/blob/master/schema/spacewalk/postgres/procs/lookup_package_capability.sql#L33
https://github.com/uyuni-project/uyuni/blob/master/backend/server/rhnServer/server_packages.py#L347-L371
https://github.com/uyuni-project/uyuni/blob/master/backend/server/importlib/backend.py#L111
https://github.com/uyuni-project/uyuni/blob/master/backend/satellite_tools/updatePackages.py#L251-L259

I understand your tests look promising, but I would not want to really make sure each and every one of these code paths (plus any that have escaped my attention) use the hash instead of the string, or perform equally well with the modification. Also, we would have to remember this particular column of this particular table works differently from all others in any future code change - whilst `INDEX`ing relevant columns as-is is the norm.

Do we have the possibility of enlarging the buffer page instead?